### ISSUE TITLE
Cannot read Kaifa E0026

### ISSUE THREAD

**Issue by Atrejoe** [nrocco/smeterd#7]
(this may be related to #6)
Using v 2.4.0 when running `smeterd read-meter -v` from my Raspbbery Pi for querying my Kaifa E0026, I get the following error message:

```
[2016-09-01 21:57:17,380] INFO New serial connection opened to /dev/ttyUSB0
[2016-09-01 21:57:17,399] INFO Start reading lines
[2016-09-01 22:04:07,813] INFO Closing connection to `/dev/ttyUSB0`.
Traceback (most recent call last):
  File "/usr/local/bin/smeterd", line 11, in <module>
    sys.exit(parse_and_run())
  File "/usr/local/lib/python3.4/dist-packages/smeterd/command.py", line 85, in parse_and_run
    args.func(args, parser=parser)
  File "/usr/local/lib/python3.4/dist-packages/smeterd/command.py", line 46, in run
    packet = meter.read_one_packet()
  File "/usr/local/lib/python3.4/dist-packages/smeterd/meter.py", line 81, in read_one_packet
    raise SmartMeterError('Received %d lines, we seem to be stuck in a loop, quitting.' % len(lines))
smeterd.meter.SmartMeterError: Received 41 lines, we seem to be stuck in a loop, quitting.
```

I'm using a P1 Converter Cable v2, I bought from https://sites.google.com/site/nta8130p1smartmeter/webshop

The USB device seems to be found (and briefly lights up when I connect it)
```
lsusb -t
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=dwc_otg/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/3p, 480M
        |__ Port 1: Dev 3, If 0, Class=Vendor Specific Class, Driver=smsc95xx, 480M
        |__ Port 2: Dev 5, If 0, Class=Vendor Specific Class, Driver=ftdi_sio, 12M
```

Any suggestions where to start looking?

**Comment by Atrejoe** [comment]
It seems I need to configure the serial port correctly, using the following python serial settings:

```
ser = serial.Serial()
ser.baudrate = 115200
ser.bytesize=serial.EIGHTBITS
ser.parity=serial.PARITY_NONE
ser.stopbits=serial.STOPBITS_ONE
ser.xonxoff=1
ser.rtscts=0
ser.timeout=20
ser.port="/dev/ttyUSB0"
```

I was able to read out the `/KFM5KAIFA-METER` meter, now I only have to figure out how to use these settings with smeterd

**Comment by Atrejoe** [comment]
When I run using medium verbosity and with the correct baud rate I get the following output:

```
~ $ smeterd -vv read-meter --baudrate 115200
[2016-09-02 19:41:20,393] INFO New serial connection opened to /dev/ttyUSB0
[2016-09-02 19:41:20,402] INFO Start reading lines
[2016-09-02 19:41:21,271] DEBUG >>
[2016-09-02 19:41:21,282] DEBUG >> /KFM5KAIFA-METER
[2016-09-02 19:41:21,288] DEBUG >>
[2016-09-02 19:41:21,298] DEBUG >> 1-3:0.2.8(42)
[2016-09-02 19:41:21,314] DEBUG >> 0-0:1.0.0(160902214119S)
[2016-09-02 19:41:21,337] DEBUG >> 0-0:96.1.1(XXXXXXXXXXXXMYSERIALHEREXXXXXXXXX)
[2016-09-02 19:41:21,352] DEBUG >> 1-0:1.8.1(000479.565*kWh)
[2016-09-02 19:41:21,368] DEBUG >> 1-0:1.8.2(000545.061*kWh)
[2016-09-02 19:41:21,383] DEBUG >> 1-0:2.8.1(000000.000*kWh)
[2016-09-02 19:41:21,398] DEBUG >> 1-0:2.8.2(000000.000*kWh)
[2016-09-02 19:41:21,410] DEBUG >> 0-0:96.14.0(0002)
[2016-09-02 19:41:21,423] DEBUG >> 1-0:1.7.0(00.268*kW
[2016-09-02 19:41:21,441] DEBUG >> 1-0:2.7.0(00.000*kW)
[2016-09-02 19:41:21,453] DEBUG >> 0-0:96.7.21(00008)
[2016-09-02 19:41:21,466] DEBUG >> 0-0:96.7.9(00003)
[2016-09-02 19:41:21,493] DEBUG >> 1-0:99.97.0(1)(0-0:96.7.19)(000101000015W)(2147483647*s)
[2016-09-02 19:41:21,506] DEBUG >> 1-0:32.32.0(00000)
[2016-09-02 19:41:21,526] DEBUG >> 1-0:52.32.0(00000)
[2016-09-02 19:41:21,559] DEBUG >> 1-0:72.32.0(00000)
[2016-09-02 19:41:21,580] DEBUG >> 1-0:32.36.0(00000)
[2016-09-02 19:41:21,596] DEBUG >> 1-0:52.36.0(00000)
[2016-09-02 19:41:21,609] DEBUG >> 1-0:72.36.0(00000)
[2016-09-02 19:41:21,619] DEBUG >> 0-0:96.13.1()
[2016-09-02 19:41:21,666] DEBUG >> 0-0:96.13.0()
[2016-09-02 19:41:21,728] DEBUG >> 1-0:31.7.0(000*A)
[2016-09-02 19:41:21,789] DEBUG >> 1-0:51.7.0(000*A)
[2016-09-02 19:41:21,852] DEBUG >> 1-0:71.7.0(000*A)
[2016-09-02 19:41:21,884] DEBUG >> 1-0:21.7.0(00.076*kW)
[2016-09-02 19:41:21,914] DEBUG >> 1-0:41.7.0(00.161*kW)
[2016-09-02 19:41:21,944] DEBUG >> 1-0:61.7.0(00.030*kW)
[2016-09-02 19:41:21,974] DEBUG >> 1-0:22.7.0(00.000*kW)
[2016-09-02 19:41:22,004] DEBUG >> 1-0:42.7.0(00.000*kW)
[2016-09-02 19:41:22,034] DEBUG >> 1-0:62.7.0(00.000*kW)
[2016-09-02 19:41:22,062] DEBUG >> 0-1:24.1.0(003)
[2016-09-02 19:41:22,109] DEBUG >> 0-1:96.1.0(XXXXXXXXXXXXMYSERIALHEREXXXXXXXXX)
[2016-09-02 19:41:22,226] DEBUG >> 0-1:24.2.1(160902210000S)(00321.031*m3)
[2016-09-02 19:41:22,233] DEBUG >> !E1BF
[2016-09-02 19:41:31,016] DEBUG >> /KFM5KAIFA-METER
[2016-09-02 19:41:31,034] DEBUG >>
[2016-09-02 19:41:31,045] DEBUG >> 1-3:0.2.8(42)
[2016-09-02 19:41:31,079] DEBUG >> 0-0:1.0.0(160902214129S)
[2016-09-02 19:41:31,094] INFO Closing connection to `/dev/ttyUSB0`.
Traceback (most recent call last):
  File "/usr/local/bin/smeterd", line 11, in <module>
    sys.exit(parse_and_run())
  File "/usr/local/lib/python3.4/dist-packages/smeterd/command.py", line 85, in parse_and_run
    args.func(args, parser=parser)
  File "/usr/local/lib/python3.4/dist-packages/smeterd/command.py", line 46, in run
    packet = meter.read_one_packet()
  File "/usr/local/lib/python3.4/dist-packages/smeterd/meter.py", line 81, in read_one_packet
    raise SmartMeterError('Received %d lines, we seem to be stuck in a loop, quitting.' % len(lines))
smeterd.meter.SmartMeterError: Received 41 lines, we seem to be stuck in a loop, quitting.
```
It looks like where my output returns !E1BF, line 78 of meter.py only checks for a line containing only a single `!`

Could my output be related to the parity setting or the byte size?

**Comment by koen01** [comment]
i'm getting more ore less similar results with my landis+gyr e350 meter.
only is see a ' !4FF5'  instead of !E1BF 



```
(smeterd) pi@raspberrypi:~/smeterd $ smeterd -vv read-meter --serial-port /dev/ttyUSB-P1 --baudrate 115200
[2016-09-04 21:10:24,500] INFO New serial connection opened to /dev/ttyUSB-P1
[2016-09-04 21:10:24,501] INFO Start reading lines
[2016-09-04 21:10:33,100] DEBUG >> /XMX5LGBBFFB231158062
[2016-09-04 21:10:33,101] DEBUG >> 
[2016-09-04 21:10:33,102] DEBUG >> 1-3:0.2.8(42)
[2016-09-04 21:10:33,103] DEBUG >> 0-0:1.0.0(160904211031S)
[2016-09-04 21:10:33,106] DEBUG >> 0-0:96.1.1(4530303034303031353732313632383134)
[2016-09-04 21:10:33,107] DEBUG >> 1-0:1.8.1(004017.711*kWh)
[2016-09-04 21:10:33,109] DEBUG >> 1-0:2.8.1(000000.000*kWh)
[2016-09-04 21:10:33,110] DEBUG >> 1-0:1.8.2(002827.154*kWh)
[2016-09-04 21:10:33,112] DEBUG >> 1-0:2.8.2(000000.000*kWh)
[2016-09-04 21:10:33,113] DEBUG >> 0-0:96.14.0(0001)
[2016-09-04 21:10:33,115] DEBUG >> 1-0:1.7.0(00.627*kW)
[2016-09-04 21:10:33,116] DEBUG >> 1-0:2.7.0(00
[2016-09-04 21:10:33,117] DEBUG >> 0-0:96.13.1()
[2016-09-04 21:10:33,132] DEBUG >> 0-0:96.13.0()
[2016-09-04 21:10:33,143] DEBUG >> 1-0:31.7.0(003*A)
[2016-09-04 21:10:33,173] DEBUG >> 1-0:21.7.0(00.627*kW)
[2016-09-04 21:10:33,204] DEBUG >> 1-0:22.7.0(00.000*kW)
[2016-09-04 21:10:33,233] DEBUG >> 0-1:24.1.0(003)
[2016-09-04 21:10:33,255] DEBUG >> 0-1:96.1.0(4730303233353631323139373231393134)
[2016-09-04 21:10:33,275] DEBUG >> 0-1:24.2.1(160904210000S)(05290.211*m3)
[2016-09-04 21:10:33,452] DEBUG >> !4FF5
[2016-09-04 21:10:42,843] DEBUG >> /XMX5LGBBFFB231158062
[2016-09-04 21:10:42,844] DEBUG >> 
[2016-09-04 21:10:42,862] DEBUG >> 1-3:0.2.8(42)
[2016-09-04 21:10:42,874] DEBUG >> 0-0:1.0.0(160904211041S)
[2016-09-04 21:10:42,896] DEBUG >> 0-0:96.1.1(4530303034303031353732313632383134)
[2016-09-04 21:10:42,904] DEBUG >> 1-0:1.8.1(004017.713*kWh)
[2016-09-04 21:10:42,914] DEBUG >> 1-0:2.8.1(000000.000*kWh)
[2016-09-04 21:10:42,924] DEBUG >> 1-0:1.8.2(002827.154*kWh)
[2016-09-04 21:10:42,934] DEBUG >> 1-0:2.8.2(000000.000*kWh)
[2016-09-04 21:10:42,953] DEBUG >> 0-0:96.14.0(0001)
[2016-09-04 21:10:42,963] DEBUG >> 1-0:1.7.0(00.631*kW)
[2016-09-04 21:10:42,973] DEBUG >> 1-0:2.7.0(00.000*kW)
[2016-09-04 21:10:43,004] DEBUG >> 0-0:96.7.21(00003)
[2016-09-04 21:10:43,014] DEBUG >> 0-0:96.7.9(00001)
[2016-09-04 21:10:43,051] DEBUG >> 1-0:99.97.0(1)(0-0:96.7.19)(151023100203S)(0000003387*s)
[2016-09-04 21:10:43,063] DEBUG >> 1-0:32.32.0(00000)
[2016-09-04 21:10:43,093] DEBUG >> 1-0:32.36.0(00000)
[2016-09-04 21:10:43,133] DEBUG >> 0-0:96.13.1()
[2016-09-04 21:10:43,153] DEBUG >> 0-0:96.13.0()
[2016-09-04 21:10:43,163] DEBUG >> 1-0:31.7.0(003*A)
[2016-09-04 21:10:43,163] INFO Closing connection to `/dev/ttyUSB-P1`.
Traceback (most recent call last):
  File "/home/pi/smeterd/bin/smeterd", line 11, in <module>
    sys.exit(parse_and_run())
  File "/home/pi/smeterd/lib/python3.4/site-packages/smeterd/command.py", line 85, in parse_and_run
    args.func(args, parser=parser)
  File "/home/pi/smeterd/lib/python3.4/site-packages/smeterd/command.py", line 46, in run
    packet = meter.read_one_packet()
  File "/home/pi/smeterd/lib/python3.4/site-packages/smeterd/meter.py", line 81, in read_one_packet
    raise SmartMeterError('Received %d lines, we seem to be stuck in a loop, quitting.' % len(lines))
smeterd.meter.SmartMeterError: Received 41 lines, we seem to be stuck in a loop, quitting.
```

**Comment by nrocco** [comment]
Thanks for reporting these things here.

Are you able to attach a sample smart meter message as a plaintext file to this thread?
Then I can add support for these meters to this library. 

-- 
Nico Di Rocco

**Comment by koen01** [comment]
see this thread on the home-assistant forum for more infos
https://community.home-assistant.io/t/support-for-reading-dutch-smart-meter-electricity-gas-p1-port/1676/16?u=koen01

**Comment by Atrejoe** [comment]
[kaifa output.txt](https://github.com/nrocco/smeterd/files/455433/kaifa.output.txt)

It seems that whatever is behing the exclamation mark is variable/random, yet always four hexadecimal digits.

**Comment by Atrejoe** [comment]
I'm working on a Windows machine, and my cable is attached to my Pi, any hints how I can build and deploy to my Pi to debug/test this, so I can make a pull request?

@koen01 you state that your gas is not returned, when looking at the regex used for total gas:
`r'^\(([0-9]{5}\.[0-9]{3})\)$'` misses the field indicator: 0-1:24.2.1, furthermore you and I have a field leading the actual value with a number followed by `S`, of which I don't known what it stands for.
The following regex does match your and my total gas volume:
` ^0-1:24\.2\.1\(\d+S\)?\(([0-9]{5}\.[0-9]{3})\*m3\)$` (meter.py, line 130)

Your measurement: `0-1:24.2.1(160904210000S)(05290.211*m3)`
My measurement: `0-1:24.2.1(160902210000S)(00321.031*m3)`

This in itself is worth a pull request I think

**Comment by nrocco** [comment]
Hi Robert,

I am currently on holidays and will only be back next week Saturday. This is definitely worth a pull request. 

The easiest is to clone this repository on your pi and run smeterd from a virtual environment so you are able to change the code, run smeterd, debug and repeat. 

See https://github.com/nrocco/smeterd/blob/master/Makefile for instructions on how to setup your development environment. 

After cloning the repository create a virtual environment by running:

    $ make develop

Now you can edit the source code and invoke smeterd like this:

    $ env/bin/smeterd 

Also notice the tests folder, it contains sample packets that I use to write unit tests:

https://github.com/nrocco/smeterd/blob/master/tests/__init__.py

You can run all tests by running

    $ make test

Hope this provides you enough info to start the PR. Otherwise I will be back end of next week. It would be great to have support for these meters in smeterd!


-- 
Nico Di Rocco

**Comment by Atrejoe** [comment]
Thanks for the info, I will use this for testing on my Pi. I guess I also should be able to run the unit tests on my Windows machine too, but I have some issues installing the dependencies, but I'll figure it out.

When possible I will include my 'telegram' and @Koen01 s telegram as unit test to prevent regression.

I would also like to officially expose the data-transfer object (`P1Packet`), so other projects (read: my Home Assistant  project) can access the details.

**Comment by Atrejoe** [comment]
@koen01 it looks like your telegram was corrupt, see line 12: `1-0:2.7.0(00.000`

This seems to bu an unclosed value for current_produced. Can you see if you have this more often?

**Comment by Atrejoe** [comment]
As @koen01 suggested I corrected the last-line-detection in 5f20592469d5fd472efb0fddaf5e3244862a51a6 (`startswith('!')`)and enlarged the maximum number of lines to read in 0b008d4777ac26f7f1ceb29a3621edca8cade822
These both changes are incorporated in #8 
With the instruction of @nrocco the code of this pull request now actually function on my pi and read the meter (yay):
```
Time                      2016-09-07 21:41:38.254541
Total kWh High consumed   560491
Total kWh Low consumed    501617
Total gas consumed        325262
Current kWh tariff        1
```

Sometimes I still have issues reading out the complete telegram. I will have to figure out if that has to do with the abovementioned serial settings

**Comment by Atrejoe** [comment]
CLosed pending pull request
